packages/labs/ssr/src/lib/declarative-style-dedupe.ts

```diff
@@ -0,0 +1,109 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {CSSResult, CSSResultOrNative} from 'lit';
+import {RenderResult} from './render-result.js';
+
+export interface Options {
+  /**
+   * To dedupe styles on your SSR'd page, the DeclarativeStyleDedupeUtility
+   * injects a global custom element onto your page. By default it is named
+   * `<lit-ssr-style-dedupe>`, but the name can be customized via the `tagName`
+   * option.
+   */
+  tagName?: string;
+}
+
+/**
+ * StyleDedupe is a utility that can be optionally passed to your SSR config to
+ * inject a small amount of JavaScript on your page.
+ *
+ * Make sure that the same instance of `DeclarativeStyleDedupeUtility` is used
+ * for a page.
+ */
+export class DeclarativeStyleDedupeUtility {
+  private hasEmittedDeclaration = false;
+  private readonly styleModuleTagName;
+  private idCounter = 0;
+  private styleToId = new Map<CSSResultOrNative, number>();
+
+  constructor(opts?: Options) {
+    this.styleModuleTagName = opts?.tagName ?? 'lit-ssr-style-dedupe';
+  }
+
+  emitCustomElementDeclaration(): string {
+    if (this.hasEmittedDeclaration) {
+      return '';
+    }
+    this.hasEmittedDeclaration = true;
+    return `
+<script>
+(function() {
+if (customElements.get('${this.styleModuleTagName}')) return;
+
+const supportsAdoptingStyleSheets = window.ShadowRoot &&
+  (window.ShadyCSS === undefined || global.ShadyCSS.nativeShadow) &&
+  'adoptedStyleSheets' in Document.prototype &&
+  'replaceSync' in CSSStyleSheet.prototype;
+
+class StyleModule extends HTMLElement {
+    static styleCache = new Map();
+
+    _cachedSS = undefined;
+    get constructableStyleSheet() {
+      if (this._cachedSS) {
+        return this._cachedSS;
+      }
+      this._cachedSS = new CSSStyleSheet();
+      this._cachedSS.replaceSync(this.children[0].textContent);
+      return this._cachedSS;
+    }
+
+    connectedCallback() {
+      const styleId = this.getAttribute('style-id');
+      const cachedStyleSheet = StyleModule.styleCache.get(styleId);
+      if (cachedStyleSheet == null) {
+        StyleModule.styleCache.set(styleId, this);
+        return;
```

Comment:
So, if the style is here, the element just uses it? That's incorrect since it won't be shared. Instead, the element should remove the style and adopt the sheet.

Reply:
Consider instead, putting the `<style>` before this element so that it's available at `connectedCallback` time. Then it can be processed here.

Reply:
Additionally, remove the de-duping element then the stylesheet is adopted so it doesn't pollute the element's DOM.

Reply:
This seems to work, see: https://lit.dev/playground/#gist=078a12b72391e1638f6cfb09391b7688

Reply:
The names need to be fixed. But `cachedStyleSheet` here is referring to the `StyleModule` that contains the style definition – the first instance of StyleModule.

If there is not already a StyleModule in the cache, then we must be the first one containing a definition.

Reply:
I edited my comment above to strike it out. I am confused as to why the first instance with the concrete `<style>` element is wrong if it isn't adopted?

I'd love to understand your very first comment better "So, if the style is here, the element just uses it? That's incorrect since it won't be shared. Instead, the element should remove the style and adopt the sheet.".

Why is it more correct to remove the style and adopt the sheet?

Reply:
We resolved at Lit Eng meeting! Thank you!